core/src/test/java/org/apache/calcite/plan/volcano/MultipleTraitConversionTest.java

```diff
@@ -0,0 +1,178 @@
+    planner.setRoot(rel);
+    planner.changeTraitsUsingConverters(rel, toTraits);
+
+    // Make sure that the equivalence set contains only the original and converted rels.
+    // It should not contain the collation enforcer, because the "from" collation already
+    // satisfies the "to" collation.
+    List<RelNode> rels = planner.getSubset(rel).set.rels;
+    assertEquals(2, rels.size());
+    assertTrue(rels.stream().anyMatch(r -> r instanceof CustomLeafRel));
+    assertTrue(rels.stream().anyMatch(r -> r instanceof CustomTraitEnforcer));
+  }
+
+  /**
+   * Leaf rel.
+   */
+  private static class CustomLeafRel extends PlannerTests.TestLeafRel {
+    CustomLeafRel(RelOptCluster cluster, RelTraitSet traits) {
+      super(cluster, traits, CustomLeafRel.class.getSimpleName());
+    }
+
+    @Override public RelNode copy(RelTraitSet traitSet, List<RelNode> inputs) {
+      return new CustomLeafRel(getCluster(), traitSet);
+    }
+
+    @Override public @Nullable RelOptCost computeSelfCost(RelOptPlanner planner,
+        RelMetadataQuery mq) {
+      return planner.getCostFactory().makeTinyCost();
+    }
+  }
+
+  /**
+   * An enforcer used by the custom trait def.
+   */
+  private static class CustomTraitEnforcer extends SingleRel {
+    private CustomTraitEnforcer(RelOptCluster cluster, RelTraitSet traits, RelNode input) {
+      super(cluster, traits, input);
+    }
+
+    @Override public RelNode copy(RelTraitSet traitSet, List<RelNode> inputs) {
+      return new CustomTraitEnforcer(getCluster(), traitSet, inputs.get(0));
+    }
+  }
+
+  /**
+   * Custom trait.
+   */
+  private static class CustomTrait implements RelTrait {
+
+    private static final CustomTrait FROM = new CustomTrait("FROM");
+    private static final CustomTrait TO = new CustomTrait("TO");
+
+    private final String label;
+
+    private CustomTrait(String label) {
+      this.label = label;
+    }
+
+    @SuppressWarnings("rawtypes")
+    @Override public RelTraitDef getTraitDef() {
+      return CustomTraitDef.INSTANCE;
+    }
+
+    @Override public boolean satisfies(RelTrait trait) {
+      return this == trait;
+    }
+
+    @Override public void register(RelOptPlanner planner) {
+      // No-op
+    }
+
+    @Override public String toString() {
+      return label;
+    }
+  }
+
+  /**
+   * Custom trait definition.
+   */
+  private static class CustomTraitDef extends RelTraitDef<CustomTrait> {
+
+    private static final CustomTraitDef INSTANCE = new CustomTraitDef();
+
+    @Override public Class<CustomTrait> getTraitClass() {
+      return CustomTrait.class;
+    }
+
+    @Override public String getSimpleName() {
+      return "custom";
+    }
+
+    @Override public @Nullable RelNode convert(
+        RelOptPlanner planner,
+        RelNode rel,
+        CustomTrait toTrait,
+        boolean allowInfiniteCostConverters
+    ) {
+      return new CustomTraitEnforcer(
+          rel.getCluster(),
+          rel.getTraitSet().plus(toTrait),
```

Comment:
I might not understand correctly:

should here be `remove existing then plus`?

Reply:
The `plus` method replaces the existing trait definitions internally. 

Before: `NONE,[0,1],FROM`
After:  `NONE,[0,1],TO`

Reply:
I see. In this case, can you also check in your unit test that for those two Rels, only the FROM and TO traits are different (others have skipped the conversion).

Reply:
This is not the goal of the test. Rels have two different traits - a collation and a custom one. But the collation of the original node satisfies the collation of the target node. Before the fix, Calcite added an unnecessary Sort on top of the original node. After the fix, Calcite enforces only the custom trait.

Reply:
Or maybe the proposal is to check traits before and after the `plus` operation?

Reply:
O I see. So the goal was to remove that Sort? 

It won't hurt though to verify 

```
Before: NONE,[0,1],FROM
After: NONE,[0,1],TO
```

Reply:
The goal was to ensure that only enforces that are really required are applied. Before the fix, the original node already has the collation that satisfies the output traits, yet, the `Sort` operator was added to MEMO.

I slightly improved the test to address your concerns: https://github.com/apache/calcite/pull/2325/commits/0aedf2e3dc2a59f1f3366acd374e349fbdb01ea4